### PR TITLE
Use proper Config when validating config properties and injecting into beans

### LINKED ISSUE #2146: Allow injecting default value of quarkus.http.port
When `quarkus.http.port` is not defined in `application.properties` and the value is injected the  deployment fails. It would be more friendly if quarkus injected default values of http port (and probably other core configuration properties).

```java
  @ConfigProperty(name = "quarkus.http.port")
  private int port;
```

```
javax.enterprise.inject.spi.DeploymentException: No config value of type [int] exists for: quarkus.http.port
	at io.quarkus.arc.runtime.ConfigDeploymentTemplate.validateConfigProperties(ConfigDeploymentTemplate.java:48)
	at io.quarkus.deployment.steps.ConfigBuildStep$validateConfigProperties7.deploy(Unknown Source)
	at io.quarkus.runner.ApplicationImpl1.doStart(Unknown Source)
	at io.quarkus.runtime.Application.start(Application.java:99)
	at io.quarkus.runner.RuntimeRunner.run(RuntimeRunner.java:119)
	at io.quarkus.test.junit.QuarkusTestExtension.doJavaStart(QuarkusTestExtension.java:236)
	at io.quarkus.test.junit.QuarkusTestExtension.createTestInstance(QuarkusTestExtension.java:301)
	at org.junit.jupiter.engine.descriptor.ClassTestDescriptor.invokeTestInstanceFactory(ClassTestDescriptor.java:299)
	at org.junit.jupiter.engine.descriptor.ClassTestDescriptor.instantiateTestClass(ClassTestDescriptor.java:289)
	at org.junit.jupiter.engine.descriptor.ClassTestDescriptor.instantiateTestClass(ClassTestDescriptor.java:281)
	at org.junit.jupiter.engine.descriptor.ClassTestDescriptor.instantiateAndPostProcessTestInstance(ClassTestDescriptor.java:269)
	at org.junit.jupiter.engine.descriptor.ClassTestDescriptor.lambda$testInstancesProvider$2(ClassTestDescriptor.java:259)
	at org.junit.jupiter.engine.descriptor.ClassTestDescriptor.lambda$testInstancesProvider$3(ClassTestDescriptor.java:263)
	at java.util.Optional.orElseGet(Optional.java:267)
	at org.junit.jupiter.engine.descriptor.ClassTestDescriptor.lambda$testInstancesProvider$4(ClassTestDescriptor.java:262)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$prepare$0(TestMethodTestDescriptor.java:98)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.prepare(TestMethodTestDescriptor.java:97)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.prepare(TestMethodTestDescriptor.java:68)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$prepare$1(NodeTestTask.java:107)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.prepare(NodeTestTask.java:107)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:75)
	at java.util.ArrayList.forEach(ArrayList.java:1257)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:38)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:125)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:135)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:123)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:122)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:80)
	at java.util.ArrayList.forEach(ArrayList.java:1257)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:38)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:125)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:135)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:123)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:122)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:80)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:32)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:51)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:170)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:154)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:90)
	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.invokeAllTests(JUnitPlatformProvider.java:142)
	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.invoke(JUnitPlatformProvider.java:117)
	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:383)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:344)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:125)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:417)
```

### PR DESCRIPTION
**Pull request by geoand**
Fixes: #2146

### CODE REVIEW

**Review by dmlloyd — DISMISSED** [review]

**Comment by geoand** [comment]
Actually the test is failing... So I will add an `In-progress` label and see what the problem is.

**Comment by geoand** [comment]
OK, so this issue goes deeper.... Although the PR gets around the runtime validation problem described in #2146, however the correct value is **not** injected into the beans.

As @mkouba said in that issue, `io.smallrye.config.inject.ConfigProducer` does:

```
    @Produces
    Config getConfig(InjectionPoint injectionPoint) {
        // return the Config for the TCCL
        return ConfigProvider.getConfig(getContextClassLoader());
    }
```

which means that we are not leveraging the Config that gets generated as part of the generated `RuntimeConfig#getRunTimeConfiguration`.

So I am not really sure what the proper approach should be here... @mkouba @dmlloyd @kenfinnigan any ideas?

**Comment by dmlloyd** [comment]
Might this method be being called before `main` runs i.e. before the configuration is loaded and installed?

**Comment by geoand** [comment]
@dmlloyd I'll debug and let you know.

**Comment by geoand** [comment]
It seems like `org.jboss.resteasy.microprofile.config.ResteasyConfigProvider#getConfig` is the first one to call `ConfigProvider#getConfig`. 
This method invocation results from `UndertowDeploymentTemplate#bootServletContainer` which is done of course in static init.

**Comment by dmlloyd** [comment]
OK that's why the run time config values are not present.  This continues the scattered discussion we've been having about the problems with Arc config injection versus static initialization: it can never be made to work unless you are explicitly injecting build time configuration.

I'll post something to the dev list I guess.

**Comment by geoand** [comment]
@dmlloyd sounds good, thanks.

**Comment by dmlloyd** [comment]
Well, on reviewing the code a bit more closely it seems that this part doesn't run until run time, so... I guess I don't know what's actually going on here.  Note that the failure message looks like this:

```
[ERROR] Failures: 
[ERROR]   PropertyInjectionTestCase.testConfigPropertyInjectedIntoTest:17 expected: <A message on 8081> but was: <A message>
[ERROR]   ServletTestCase.testServlet:33 1 expectation failed.
Response body doesn't match expectation.
Expected: is "A message"
  Actual: A message on null
```

That's two failures.  The first one looks like it didn't write the correct message at all - I mean "on" is missing so the output it expected is not the one that was produced by your change.  Then the second has the opposite problem.  Maybe the test is bad?

**Comment by geoand** [comment]
@dmlloyd Indeed, the test is bad (good catch!), but the problem still exists even when the test is corrected.

**Comment by dmlloyd** [comment]
OK that makes a little more sense. :)

That said I don't think the test actually tests the original problem.  It tests that a run time config property is available to a static init CDI bean which will never work, so the test will never pass.

**Comment by geoand** [comment]
Test updated to inject into a CDI bean - still fails :)

```
[ERROR] Failures: 
[ERROR]   JaxRSTestCase.testConfigInjection:46 1 expectation failed.
Response body doesn't match expectation.
Expected: is "8081"
  Actual: 

```

**Comment by dmlloyd** [comment]
Are you certain that the `ServiceWithQuarkusConfig` is instantiated at run time?  If yes, then there's a bug (somewhere) and it might be related to the original issue.  If no, then there's a systemic problem in that run time configuration injection can seemingly never be available to things that use REST.

**Comment by geoand** [comment]
The latest push should make things clear:

Injecting a value that has been configured in `application.properties` works as expected. Injection `quarkus.http.port` does not (my best guess is what I explained above with the ConfigProvider stuff).

**Comment by geoand** [comment]
@mkouba is there a CDI way we could ignore `io.smallrye.config.inject.ConfigProducer` and instead provide a "producer" bean of our own?

**Comment by geoand** [comment]
@dmlloyd @mkouba I believe that my latest change should fix the issue.

File: extensions/arc/runtime/src/main/java/io/quarkus/arc/runtime/CustomConfigProducer.java [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+import org.eclipse.microprofile.config.spi.ConfigProviderResolver;
+
+import io.smallrye.config.inject.ConfigProducerUtil;
+
+/**
+ * This class is the same as io.smallrye.config.inject.ConfigProducer
+ * but uses the proper Quarkus way of obtaining org.eclipse.microprofile.config.Config
+ */
+@ApplicationScoped
+public class CustomConfigProducer {
```
Comment (gsmet) [inline extensions/arc/runtime/src/main/java/io/quarkus/arc/runtime/CustomConfigProducer.java]: Maybe it should be called `QuarkusConfigProducer`?

Comment (geoand) [reply extensions/arc/runtime/src/main/java/io/quarkus/arc/runtime/CustomConfigProducer.java]: That's a good idea, I'll fix it

Comment (geoand) [reply extensions/arc/runtime/src/main/java/io/quarkus/arc/runtime/CustomConfigProducer.java]: Done

**Review by dmlloyd — APPROVED** [review]
OK looks good to me.  But it's just a workaround until we have the right configuration setup procedure.  I'll see if I can resurrect that branch today (it predates a couple big refactors so it might take some time).

**Comment by geoand** [comment]
@dmlloyd sounds good. Should I merge this in the mean time?

**Comment by dmlloyd** [comment]
Yes, let's merge it.

**Comment by geoand** [comment]
Cool, thanks!

**Pull request merged**